docs/data/data-grid/server-side-data/tree-data.md

```diff
@@ -2,14 +2,80 @@
 title: React Server-side tree data
 ---
 
-# Data Grid - Server-side tree data [<span class="plan-pro"></span>](/x/introduction/licensing/#pro-plan 'Pro plan')🚧
+# Data Grid - Server-side tree data [<span class="plan-pro"></span>](/x/introduction/licensing/#pro-plan 'Pro plan')
 
 <p class="description">Tree data lazy-loading with server-side data source.</p>
 
-:::warning
-This feature isn't implemented yet. It's coming.
+To dynamically load tree data from the server, including lazy-loading of children, you must create a data source and pass the `unstable_dataSource` prop to the Data Grid, as detailed in the [overview section](/x/react-data-grid/server-side-data/).
 
-👍 Upvote [issue #3377](https://github.com/mui/mui-x/issues/3377) if you want to see it land faster.
+Additionally, you must supply the following required props, listed and explained below.
 
-Don't hesitate to leave a comment on the same issue to influence what gets built. Especially if you already have a use case for this component, or if you are facing a pain point with the [currently proposed workaround](https://mui.com/x/react-data-grid/tree-data/#children-lazy-loading).
+```tsx
+<DataGridPro
+  {...data}
+  unstable_dataSource={dataSource}
+  getGroupKey={getGroupKey}
+  hasChildren={hasChildren}
+  getChildrenCount={getChildrenCount}
```

Comment:
Do you think we can move these props into the data source itself? Would there be any downsides to this?
It feels like they belong to the data source, and won't be used without data source.

Reply:
> Would there be any downsides to this?

Not any particular, I just had an impression that actual row data and metadata should be separate, the datasource interface should just be a representation of the actual datasource (BE/database/etc).

The meta information and other props like `columns`, `getGroupKey`, `hasChildren`, `serverSideCache`, `defaultGroupingExpansionDepth`, etc. should be part of grid props, so that if we had to compute them separately (using useMemo for example) we don't have to update the reference of `unstable_dataSource` (which may result in resetting the whole state and data refetch)

This is my personal opinion though, I'm not sure if we'll ever need to update the references of `getGroupKey`, `hasChildren`, `getChildrenCount` props. So maybe we _can_ make them part of the data source interface. Let me know if you feel it to be better that way and I'll refactor it.

Reply:
> I just had an impression that actual row data and metadata should be separate, the datasource interface should just be a representation of the actual datasource (BE/database/etc).

For `hasChildren` and `getChildrenCount` specifically, I think they really belong to the data source since you'll be querying your backend/db in those methods, similar to `getRows`.

When deciding whether something should be part of the `dataSource` interface, it might be helpful to look at it from an encapsulation point of view – if you extract a reusable data source, what do you need to define there for the data source to be usable?. `hasChildren` and `getChildrenCount` seem to fall in this bucket, similar to `getGroupKey`.

What do you think?